File: apps/web/src/features/experiments/api/controlSets.ts
import type { ControlSet } from "@pathfinder/shared";
import { requestJson } from "@/lib/api/http";

export async function listControlSets(siteId: string): Promise<ControlSet[]> {
  return await requestJson<ControlSet[]>("/api/v1/control-sets", {
    query: { siteId },
  });
}

export async function getControlSet(id: string): Promise<ControlSet> {
  return await requestJson<ControlSet>(`/api/v1/control-sets/${id}`);
}

export async function createControlSet(body: {
  name: string;
  siteId: string;
  recordType: string;
  positiveIds: string[];
  negativeIds: string[];
  source?: string;
  tags?: string[];
  provenanceNotes?: string;
  isPublic?: boolean;
}): Promise<ControlSet> {
  return await requestJson<ControlSet>("/api/v1/control-sets", {
    method: "POST",
    body,
  });
}

export async function deleteControlSet(id: string): Promise<void> {
  await requestJson(`/api/v1/control-sets/${id}`, { method: "DELETE" });
}

export async function getExperimentReport(experimentId: string): Promise<void> {
  const resp = await fetch(`/api/v1/experiments/${experimentId}/report`, {
    credentials: "include",
  });
  if (!resp.ok) throw new Error(`Report generation failed: ${resp.status}`);
  const blob = await resp.blob();
  const a = document.createElement("a");
  a.href = URL.createObjectURL(blob);
  a.download = `experiment-report.html`;
  a.click();
  URL.revokeObjectURL(a.href);
}

File: apps/api/src/pathfinder/transport/http/routers/control_sets.py
"""CRUD endpoints for reusable control gene sets."""

from uuid import UUID

from assistant_core.platform.pydantic_base import CamelModel
from fastapi import APIRouter
from pydantic import Field

from pathfinder.platform.errors import ValidationError as CoreValidationError
from pathfinder.services.control_sets import (
    ControlSetResponse,
    ControlSetService,
    NewControlSet,
)
from pathfinder.transport.http.deps import CurrentUser, DBSession, SiteIdQuery
from pathfinder.transport.http.schemas.site_id import SiteId

router = APIRouter(prefix="/api/v1/control-sets", tags=["control-sets"])


class CreateControlSetRequest(CamelModel):
    """Payload for creating a new control set."""

    name: str = Field(min_length=1, max_length=255)
    site_id: SiteId
    record_type: str = Field(min_length=1, max_length=100)
    positive_ids: list[str] = Field(default_factory=list)
    negative_ids: list[str] = Field(default_factory=list)
    source: str | None = Field(None, max_length=50)
    tags: list[str] = Field(default_factory=list)
    provenance_notes: str | None = Field(None)
    is_public: bool = Field(default=False)


@router.get("", response_model=list[ControlSetResponse])
async def list_control_sets(
    session: DBSession,
    user_id: CurrentUser,
    site_id: SiteIdQuery = None,
    tags: str | None = None,
) -> list[ControlSetResponse]:
    """List control sets visible to the current user."""
    tag_list = [t.strip() for t in tags.split(",") if t.strip()] if tags else None
    if not site_id:
        raise CoreValidationError(
            title="Missing required parameter",
            detail="siteId query parameter is required",
        )
    return await ControlSetService(session).list_for_site(
        site_id=site_id,
        user_id=user_id,
        tags=tag_list,
    )


@router.get("/{control_set_id}", response_model=ControlSetResponse)
async def get_control_set(
    control_set_id: UUID,
    session: DBSession,
    user_id: CurrentUser,
) -> ControlSetResponse:
    """Get a single control set by ID."""
    return await ControlSetService(session).get(control_set_id, user_id)


@router.post("", response_model=ControlSetResponse, status_code=201)
async def create_control_set(
    body: CreateControlSetRequest,
    session: DBSession,
    user_id: CurrentUser,
) -> ControlSetResponse:
    """Create a new control set."""
    spec = NewControlSet(
        name=body.name,
        site_id=body.site_id,
        record_type=body.record_type,
        positive_ids=body.positive_ids,
        negative_ids=body.negative_ids,
        source=body.source,
        tags=body.tags,
        provenance_notes=body.provenance_notes,
        is_public=body.is_public,
    )
    return await ControlSetService(session).create(spec, user_id=user_id)


@router.delete("/{control_set_id}", status_code=204)
async def delete_control_set(
    control_set_id: UUID,
    session: DBSession,
    user_id: CurrentUser,
) -> None:
    """Delete a control set owned by the current user."""
    await ControlSetService(session).delete(control_set_id, user_id)
